src/Compilers/Core/Portable/EmbeddedText.cs

```diff
@@ -0,0 +1,345 @@
+            Checksum = checksum;
+            ChecksumAlgorithm = checksumAlgorithm;
+            Blob = blob;
+        }
+
+        /// <summary>
+        /// Creates the blob to be saved to the PDB.
+        /// </summary>
+        internal static ImmutableArray<byte> CreateBlob(Stream stream)
+        {
+            Debug.Assert(stream != null);
+            Debug.Assert(stream.CanRead);
+            Debug.Assert(stream.CanSeek);;
+            Debug.Assert(stream.Length <= int.MaxValue);
+
+            stream.Seek(0, SeekOrigin.Begin);
+            int length = (int)stream.Length;
+
+            if (length < CompressionThreshold)
+            {
+                var builder = Cci.PooledBlobBuilder.GetInstance();
+                builder.WriteInt32(0);
+
+                int bytesWritten = builder.TryWriteBytes(stream, length);
+                if (length != bytesWritten)
+                {
+                    throw new EndOfStreamException();
+                }
+
+                return builder.ToImmutableArrayAndFree();
+            }
+            else
+            {
+                Debug.Assert(length > 0);
+                var builder = BlobBuildingStream.GetInstance();
+                builder.WriteInt32(length);
+
+                using (var deflater = new DeflateStream(builder, CompressionLevel.Optimal))
+                {
+                    stream.CopyTo(deflater);
+                }
+
+                return builder.ToImmutableArrayAndFree();
+            }
+        }
+
+        internal static ImmutableArray<byte> CreateBlob(byte[] buffer, int length)
+        {
+            Debug.Assert(buffer != null);
+            Debug.Assert(length > 0);
+            Debug.Assert(length <= buffer.Length);
+
+            if (length < CompressionThreshold)
+            {
+                var builder = Cci.PooledBlobBuilder.GetInstance();
+                builder.WriteInt32(0);
+                builder.WriteBytes(buffer, 0, length);
+                return builder.ToImmutableArrayAndFree();
+            }
+            else
+            {
+                Debug.Assert(length > 0);
+                var builder = BlobBuildingStream.GetInstance();
+                builder.WriteInt32(length);
+
+                using (var deflater = new DeflateStream(builder, CompressionLevel.Optimal))
+                {
+                    deflater.Write(buffer, 0, length);
+                }
+
+                return builder.ToImmutableArrayAndFree();
+            }
+        }
+
+        private static ImmutableArray<byte> CreateBlob(SourceText sourceText)
+        {
+            Debug.Assert(sourceText != null);
+            Debug.Assert(sourceText.CanBeEmbedded);
+            Debug.Assert(sourceText.Encoding != null);
+            Debug.Assert(sourceText.PrecomputedEmbeddedTextBlob.IsDefault);
+
+            int maxByteCount = int.MaxValue;
+            try { maxByteCount = sourceText.Encoding.GetMaxByteCount(sourceText.Length);  }
```

Comment:
Also add a comment why we are swallowing the exception.

---
In reply to: [72831841](https://github.com/dotnet/roslyn/pull/12353#discussion_r72831841) [](ancestors = 72831841)